### PR TITLE
Add an 'IconNames' property to each device

### PR DESCRIPTION
**Pull request by hughsie**
This allows us to show the devices in a GUI with a nice icon. Some of the icon
mappings are not perfect and I'll be asking the GNOME designers for some
additions to the icon specification.

### CODE REVIEW

**Comment by ybernat** [comment]
For Thunderbolt,there is the official icon (see here: https://thunderbolttechnology.net/), but I assume there is a licensing issue...
Do you have any suggestion? What are your terms for including it here?

**Comment by hughsie** [comment]
This isn't really a "show branding" icon, more a "kind of device" icon -- it's also got to be a stock icon rather than anything custom.

**Comment by ybernat** [comment]
Oh, I see.
So we have to look for the PCI devices connected behind TBT if we want to figure out the actual device type, probably.

**Comment by hughsie** [comment]
I think, if that's possible, that would be great.

**Comment by ybernat** [comment]
I found this: https://wiki.debian.org/HowToIdentifyADevice/PCI
The point is that as long as the device isn't authorized, the PCI devices aren't connected (which is the whole point of the authorization). Maybe an "unknown" (question mark would fit) for this case and update when the device changed (the plugin already watches this event).

**Comment by hughsie** [comment]
Maybe this should be `icons` rather than `icon_names` -- the latter is very verbose and I'm not sure the distrinction between `GIcons` and stock icon names is useful in an API that's never going to be passing pixbufs around.

**Comment by superm1** [comment]
>This allows us to show the devices in a GUI with a nice icon. Some of the icon
>mappings are not perfect and I'll be asking the GNOME designers for some
>additions to the icon specification.

I think it would make sense to have a Thunderbolt specific icon if they don't already have this planned.  It will (presumably) be needed for when GUI TBT security level support too.

>I found this: https://wiki.debian.org/HowToIdentifyADevice/PCI
>The point is that as long as the device isn't authorized, the PCI devices aren't connected (which is the >whole point of the authorization). Maybe an "unknown" (question mark would fit) for this case and >update when the device changed (the plugin already watches this event).

At least from a fwupd perspective it doesn't matter that it's PCIe devices behind the TBT bridge.  What you're showing in fwupd the NVM for and updating is the TBT controller itself.  I think it would be best to use a thunderbolt icon for all of these: host and device.

File: libfwupd/fwupd-device.c [hunk 1]
```diff
@@ -254,6 +255,70 @@ fwupd_device_get_guid_default (FwupdDevice *device)
...
+
+	g_return_val_if_fail (FWUPD_IS_DEVICE (device), FALSE);
+
+	for (i = 0; i < priv->icon_names->len; i++) {
+		const gchar *icon_name_tmp = g_ptr_array_index (priv->icon_names, i);
+		if (g_strcmp0 (icon_name, icon_name_tmp) == 0)
+			return TRUE;
+	}
+	return FALSE;
+}
```
Comment (superm1) [inline libfwupd/fwupd-device.c]: Is there a good reason to have `fwupd_device_has_icon_name` available outside of libfwupd?  It just seems to me it's better to keep this code as part of the function `fwupd_device_add_icon_name ` and then not expose it elsewhere.

**Comment by ybernat** [comment]
> At least from a fwupd perspective it doesn't matter that it's PCIe devices behind the TBT bridge. What you're showing in fwupd the NVM for and updating is the TBT controller itself. I think it would be best to use a thunderbolt icon for all of these: host and device.

This is what I thought in first place, but from @hughsie comment I understood he prefers showing the device functionality (disk, audio, network, all of them together for a dock, etc.), which seems like the whole point of allowing an array of icons instead of a single one. So maybe we want both, TBT icon and icons for the actual functionality it implements.

File: libfwupd/fwupd-device.c (hunk 1 above)
Comment (hughsie) [reply libfwupd/fwupd-device.c]: Agree.

**Comment by hughsie** [comment]
> I think it would make sense to have a Thunderbolt specific icon

if it's an icon (rather than an icon_*name*) we could make it a stock name by default, and also specify it can be a full path on the filesystem.

File: libfwupd/fwupd-device.c [hunk 2]
```diff
@@ -254,6 +255,58 @@ fwupd_device_get_guid_default (FwupdDevice *device)
...
+ * Adds the icon name if it does not already exist.
+ *
+ * Since: 0.9.8
+ **/
+void
+fwupd_device_add_icon (FwupdDevice *device, const gchar *icon)
+{
+	FwupdDevicePrivate *priv = GET_PRIVATE (device);
+	g_return_if_fail (FWUPD_IS_DEVICE (device));
+	if (fwupd_device_has_icon (device, icon))
```
Comment (superm1) [inline libfwupd/fwupd-device.c]: Weren't you going to remove this (and move the code in)?

Comment (hughsie) [reply libfwupd/fwupd-device.c]: It's unexported and static now.

**Pull request merged**